### PR TITLE
Improve suggested pairings and add memo suggestions using GPT-4

### PR DESCRIPTION
**Pull request by YodaLightsabr**
## Summary of the problem
Suggested pairings currently rely on janky regular expressions to extract features from receipts. People assume that suggested pairings are accurate even if there are errors.

## Describe your changes
To fix this, suggested pairings will now use GPT-4 to extract features. The extracted features will also be used for suggesting memos to the user.

image
<small>The pairing suggestion screen inside of Receipt Bin</small>

image
<small>A suggested memo on the HCB code page</small>

### CODE REVIEW

File: app/models/receipt.rb [hunk 1]
```diff
@@ -4,14 +4,21 @@
...
-#  receiptable_type           :string
-#  textual_content_ciphertext :text
-#  upload_method              :integer
-#  created_at                 :datetime         not null
-#  updated_at                 :datetime         not null
-#  receiptable_id             :bigint
-#  user_id                    :bigint
+#  id                              :bigint           not null, primary key
+#  receiptable_type                :string
+#  suggested_amount_cents_subtotal :integer
```
Comment (garyhtou) [inline app/models/receipt.rb]: Let's rename all the new columns to `extracted_*` (except for `suggested_memo`)

Comment (garyhtou) [inline app/models/receipt.rb]: Rename this to `extracted_subtotal_amount_cents`

File: app/models/receipt.rb [hunk 2]
```diff
@@ -4,14 +4,21 @@
...
+#  suggested_amount_cents_total    :integer
```
Comment (garyhtou) [inline app/models/receipt.rb]: total amount cents

File: app/services/receipt_service/extract.rb [hunk 3]
```diff
@@ -10,95 +10,73 @@ def run!
...
-          before_fragment:,
-          match:,
-          position:,
-          after_fragment:
+      conn = Faraday.new(
+        url: "https://api.openai.com",
+        headers: {
+          "Content-Type"  => "application/json",
+          "Authorization" => "Bearer #{Rails.application.credentials.openai.api_key}",
+          "OpenAI-Beta"   => "assistants=v1"
```
Comment (garyhtou) [inline app/services/receipt_service/extract.rb]: is this still needed?

File: app/services/receipt_service/extract.rb [hunk 4]
```diff
@@ -10,95 +10,73 @@ def run!
...
+              role: "system",
+              content: prompt
+            },
+            {
+              role: "user",
+              content: @textual_content
+            }
+          ]
+        }.to_json
       end
```
Comment (garyhtou) [inline app/services/receipt_service/extract.rb]:
```suggestion
      response = conn.post("/v1/chat/completions", {
        model: "gpt-4",
        messages: [
          {
            role: "system",
            content: prompt
          },
          {
            role: "user",
            content: @textual_content
          }
        ]
      })
```

File: app/services/receipt_service/extract.rb [hunk 5]
```diff
@@ -10,95 +10,73 @@ def run!
       @textual_content = @receipt.textual_content || @receipt.extract_textual_content!
       return nil if @textual_content.nil?
```
Comment (garyhtou) [inline app/services/receipt_service/extract.rb]:
```suggestion
      return nil if @textual_content.blank?
```

File: app/services/receipt_service/extract.rb [hunk 6]
```diff
@@ -10,95 +10,73 @@ def run!
...
-          position:,
-          after_fragment:
+      conn = Faraday.new(
+        url: "https://api.openai.com",
+        headers: {
+          "Content-Type"  => "application/json",
+          "Authorization" => "Bearer #{Rails.application.credentials.openai.api_key}",
+          "OpenAI-Beta"   => "assistants=v1"
         }
+      )
```
Comment (garyhtou) [inline app/services/receipt_service/extract.rb]: Give it a block and configure it with a json middleware for the response.

https://github.com/hackclub/bank/blob/dec9f072132d7c41bcb6d57d948f2dc986d12341/app/services/column_service.rb#L23-L28

we should also have it raise an error (configure the `:raise_error` middleware)

File: app/models/receipt.rb [hunk 7]
```diff
@@ -4,14 +4,21 @@
...
+#  suggested_card_last4            :string
```
Comment (garyhtou) [inline app/models/receipt.rb]: last 4 needs to be encrypted

File: app/services/receipt_service/extract.rb [hunk 8]
```diff
@@ -10,95 +10,73 @@ def run!
...
-          [month, day, year],
-          [day, month, year],
-          [year, month, day]
-        ]
-      end.flatten(1).reject do |date|
-        month, day, year = date
-
-        month > 12 || day > 31 || year < 1000 || year > Time.now.year + 1
-      end
+      @receipt.update(
```
Comment (garyhtou) [inline app/services/receipt_service/extract.rb]:
```suggestion
      @receipt.update!(
```

File: app/services/receipt_service/suggest.rb [hunk 9]
```diff
@@ -12,32 +12,47 @@ def run!(include_details: false)
...
-    def self.weights
-      {
-        amount_cents: 1,
-        date: 1000,
-        card_last_four: 1000,
-        merchant_zip_code: 500,
-        merchant_city: 500,
-        merchant_phone: 500,
-        merchant_name: 500
+    def distance(txn)
```
Comment (garyhtou) [inline app/services/receipt_service/suggest.rb]: rename `txn` to `hcb_code`

File: app/services/receipt_service/suggest.rb [hunk 10]
```diff
@@ -12,32 +12,47 @@ def run!(include_details: false)
...
+      return if @extracted.nil?
+
+      distances = {
+        amount_cents: {
+          value: begin
+            diff = (txn.amount_cents - @extracted.amount_cents).abs
+            diff == 0 ? 0 : [10 * Math.log((diff + 1) / 3.0) + 30, 20].max
+          end,
+          weight: 100,
+        },
```
Comment (garyhtou) [inline app/services/receipt_service/suggest.rb]: We have a couple of goals with this formula. Please tweak the formula if necessary, and leave comments explaining our goals and what the formula does.

### Goals
- Exact match has a distance of 0
- The range of `value` should be well-defined
	- This means that we must know what the min and max are for this variable
- A $1 difference for a $1 million HcbCode, should be given a smaller distance than a $1 difference for a $10 HcbCode.

**Suggestion:** The formula should be calculating "% wrong"

File: app/services/receipt_service/extract.rb [hunk 11]
```diff
@@ -10,95 +10,73 @@ def run!
...
+        suggested_amount_cents_subtotal: data.amount_cents_subtotal&.to_i,
+        suggested_amount_cents_total: data.amount_cents_total&.to_i,
+        suggested_card_last4: data.card_last_four,
+        suggested_date: data.date.to_date,
+        suggested_memo: data.memo,
+        suggested_merchant_name: data.merchant_name,
+        suggested_merchant_url: data.merchant_url,
+        suggested_merchant_zip_code: data.merchant_zip_code
```
Comment (garyhtou) [inline app/services/receipt_service/extract.rb]: strip these values if we're storing as a string

File: app/services/receipt_service/suggest.rb [hunk 12]
```diff
@@ -12,32 +12,47 @@ def run!(include_details: false)
...
+        card_last_four: {
+          value: @extracted.card_last_four == txn.card&.last4,
+          weight: 100,
+        },
+        date: {
+          value: @extracted.date.to_date == txn.date,
+          weight: 100,
+        },
+        merchant_zip_code: {
+          value: @extracted.merchant_zip_code == txn.postal_code,
```
Comment (garyhtou) [inline app/services/receipt_service/suggest.rb]: `postal_code` is not defined on HcbCode. maybe you need to go through the stripe merchant first?

File: app/services/receipt_service/suggest.rb [hunk 13]
```diff
@@ -12,32 +12,47 @@ def run!(include_details: false)
...
+          weight: 50,
+        },
+        merchant_name: {
+          value: @extracted.merchant_name == txn.stripe_merchant.name,
```
Comment (garyhtou) [inline app/services/receipt_service/suggest.rb]: notes from Ian:
```
@extracted.merchant_zip_code == txn.stripe_merchant[:postal_code],
          weight: 50,
        },
        merchant_name: {
          value: @extracted.merchant_name == txn.stripe_merchant[:name],
```

File: app/services/receipt_service/suggest.rb [hunk 14]
```diff
@@ -12,32 +12,47 @@ def run!(include_details: false)
...
+      distances = {
+        amount_cents: {
+          value: begin
+            diff = (txn.amount_cents - @extracted.amount_cents).abs
+            diff == 0 ? 0 : [10 * Math.log((diff + 1) / 3.0) + 30, 20].max
+          end,
+          weight: 100,
+        },
+        card_last_four: {
+          value: @extracted.card_last_four == txn.card&.last4,
```
Comment (garyhtou) [inline app/services/receipt_service/suggest.rb]:
```suggestion
          value: @extracted.card_last_four == txn.card&.last4 ? 1 : 0,
```

File: db/schema.rb [hunk 15]
```diff
@@ -1210,6 +1210,16 @@
...
+  create_table "metrics", force: :cascade do |t|
+    t.string "type", null: false
+    t.jsonb "metric"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.string "subject_type"
+    t.bigint "subject_id"
+    t.index ["subject_type", "subject_id"], name: "index_metrics_on_subject"
+  end
+
```
Comment (garyhtou) [inline db/schema.rb]: unrelated

File: app/services/receipt_service/suggest.rb (hunk 13 above)
Comment (garyhtou) [inline app/services/receipt_service/suggest.rb]: If zip code from stripe is `00000`, then say that it's `nil` (we don't have it).

https://www.mapofzipcodes.com/blog/00000-zip-code
^ add this link in the code too and mention that 7% of our postal code in prod are `0000`

**Comment by sampoder** [comment]
https://github.com/hackclub/hcb/pull/6048 has additional context about the RTesseract changes

File: app/models/hcb_code.rb [hunk 16]
```diff
@@ -535,4 +535,10 @@ def accepts_receipts?
     !reimbursement_expense_payout?
   end
 
+  def suggested_memos
+    receipts.map do |receipt|
+      receipt.suggested_memo
+    end.compact
```
Comment (garyhtou) [inline app/models/hcb_code.rb]:
```suggestion
    receipts.pluck(:suggested_memo).compact
```

File: app/models/user.rb [hunk 17]
```diff
@@ -9,6 +9,7 @@
 #  birthday_ciphertext      :text
 #  comment_notifications    :integer          default("all_threads"), not null
 #  email                    :text
+#  enabled_notifications    :integer          default(0), not null
```
Comment (YodaLightsabr) [inline app/models/user.rb]:
```suggestion
```

File: app/views/hcb_codes/_stripe_card.html.erb [hunk 18]
```diff
@@ -34,7 +34,21 @@
...
+            aria-label="Use AI to generate a memo"
+            data-controller="navigation"
+            data-action="click->navigation#navigate"
+            data-navigation-location-param="<%= "#{edit_hcb_code_path(@hcb_code)}?display_ai_memo=true" %>"
+            data-navigation-frame-param="<%= dom_id(@hcb_code) %>">
+            <%= inline_icon "rep", size: 36, style: "color: #a633d6; margin: -10px 0px; transform: translateY(-2px);" %>
+          </span>
+        <% end %>
+
+        <%# TODO: look into using this: render "hcb_codes/memo_stream", hcb_code: @hcb_code %>
```
Comment (sampoder) [inline app/views/hcb_codes/_stripe_card.html.erb]: We'll want to keep this, https://github.com/hackclub/hcb/pull/6265 relies on it. Basically what it does is ensure that every memo on the page for this HCB code is the same.

**Comment by sampoder** [comment]
This looks pretty good to me other than that small thing with the stream!

File: app/views/receipts/_receipt.html.erb [hunk 19]
```diff
@@ -101,14 +102,97 @@
...
+          <div class="fs-mask" style="grid-template-columns: 7rem auto;">
+            <strong style="font-weight: 600;">Uploaded</strong>
+            <%= relative_timestamp receipt.created_at %>
+          </div>
+
+          <div class="fs-mask" style="grid-template-columns: 7rem auto;">
+            <strong style="font-weight: 600;">Size</strong>
+            <%= number_to_human_size receipt.file.blob.byte_size %>
+          </div>
+        </section>
```
Comment (sampoder) [inline app/views/receipts/_receipt.html.erb]:
```suggestion
        <section class="details mt1 break-words">
          <div class="fs-mask" style="grid-template-columns: 7rem calc(100% - 7rem);">
            <strong style="font-weight: 600;">Name</strong>
            <%= receipt.file.blob.filename %>
          </div>

          <div class="fs-mask" style="grid-template-columns: 7rem auto;">
            <strong style="font-weight: 600;">Uploaded</strong>
            <%= relative_timestamp receipt.created_at %>
          </div>

          <div class="fs-mask" style="grid-template-columns: 7rem auto;">
            <strong style="font-weight: 600;">Size</strong>
            <%= number_to_human_size receipt.file.blob.byte_size %>
          </div>
        </section>
```

These changes prevent overflow like:

Screenshot 2024-06-08 at 3 32 43 PM

Instead it looks like:

Screenshot 2024-06-08 at 3 58 40 PM

There might be a better way to do this, I'm not sure.

File: app/views/receipts/_receipt.html.erb [hunk 20]
```diff
@@ -101,14 +102,97 @@
...
+        </div>
+      <% end %>
+      <div style="width: 20rem; padding: 1.5rem; border-radius: 0rem 0.75rem 0.75rem 0rem; display: flex; flex-direction: column;" class="bg-white dark:bg-dark">
+        <h2 class="mt-0">
+          <span style="vertical-align: text-bottom;"><%= user_mention(receipt.user, { avatar: { style: "vertical-align: baseline;" }, style: "color: red;" }) %></span>'s receipt
+        </h2>
+
+        <% if receipt.suggested_memo %>
+          <p class="mt-0"><%= receipt.suggested_memo %></p>
+        <% end %>
```
Comment (sampoder) [inline app/views/receipts/_receipt.html.erb]: I feel like this needs some context, it feels a bit random at the moment.

Screenshot 2024-06-08 at 4 00 42 PM

File: app/assets/javascripts/ui.js [hunk 21]
```diff
@@ -504,6 +504,15 @@ document.addEventListener("turbo:before-stream-render", ((event) => {
           element.reload();
         }
       });
+    } else if (streamElement.action == "refresh_suggested_pairings") {
+      const turboStreamElements = document.querySelectorAll('turbo-frame');
+      turboStreamElements.forEach(element => {
+        if (element.id.startsWith("suggested_pairings")) {
+          console.log("reloading!")
```
Comment (YodaLightsabr) [inline app/assets/javascripts/ui.js]:
```suggestion
```

File: app/assets/stylesheets/components/_modals.scss [hunk 22]
```diff
@@ -85,8 +85,8 @@
     }
 
     &.modal--huge {
-      width: 58rem;
-      max-width: 95vw;
+      width: 70rem;
+      max-width: 90vw;
     }
```
Comment (YodaLightsabr) [inline app/assets/stylesheets/components/_modals.scss]: Add support for addtl. receipt details for mobile screens

File: app/controllers/receipts_controller.rb [hunk 23]
```diff
@@ -169,7 +169,9 @@ def create
     notify_airbrake(e)
 
     flash_type = :error
-    flash_message = e.message
+    flash_message = "There was an error uploading your receipt"
+
+    puts e.message # TODO: Remove during code review. This exists because sometimes errors are too long to be stored with a cookie with flash.
```
Comment (YodaLightsabr) [inline app/controllers/receipts_controller.rb]:
```suggestion
    flash_message = "There was an error uploading your receipt. Please try again."
```

File: app/controllers/static_pages_controller.rb [hunk 24]
```diff
@@ -175,6 +175,13 @@ def my_inbox
     end
   end
 
+  def suggested_pairings
+    render partial: "static_pages/suggested_pairings", locals: {
+      pairings: current_user.receipt_bin.suggested_receipt_pairings,
+      current_slide: 0
```
Comment (YodaLightsabr) [inline app/controllers/static_pages_controller.rb]: Check current slide if you've already gone through slides & the turbo frame is reloaded.

File: app/services/receipt_service/suggest.rb [hunk 25]
```diff
@@ -51,108 +108,50 @@ def best_match
...
     end
 
-    def safe_date(month, day, year)
-      begin
-        Date.new(year, month, day)
-      rescue Date::Error => e
-        nil
-      end
-    end
+    # def distance(txn)
```
Comment (YodaLightsabr) [inline app/services/receipt_service/suggest.rb]: Remove this and other comments that are no longer needed

File: app/views/hcb_codes/_ai_memo.html.erb [hunk 26]
```diff
@@ -0,0 +1,13 @@
+<span id="ai_memo">
+  <%= turbo_stream_from hcb_code, "ai_memo" %>
+  <% if hcb_code.suggested_memos.any? && hcb_code.custom_memo.nil? %>
+    <span class="pointer tooltipped tooltipped--w"
+      aria-label="Use AI to generate a memo"
+      data-controller="navigation"
+      data-action="click->navigation#navigate"
+      data-navigation-location-param="<%= "#{edit_hcb_code_path(hcb_code)}?display_ai_memo=true" %>"
+      data-navigation-frame-param="<%= dom_id(hcb_code) %>">
+      <%= inline_icon "rep", size: 36, style: "color: #a633d6; margin: -10px 0px; transform: translateY(-2px);" %>
```
Comment (YodaLightsabr) [inline app/views/hcb_codes/_ai_memo.html.erb]:
```suggestion
      <%= inline_icon "rep", size: 36, style: "color: var(--ai); margin: -10px 0px; transform: translateY(-2px);" %>
```

File: app/views/receipts/_receipt.html.erb (hunk 20 above)
Comment (YodaLightsabr) [reply app/views/receipts/_receipt.html.erb]: More specific suggestion: swap Merchant/Amount/Data with Name/Uploaded/Size

File: app/views/static_pages/_suggested_pairings.html.erb [hunk 27]
```diff
@@ -45,46 +45,67 @@
                   <tbody data-behavior="transactions">
                     <% if pairing.hcb_code.canonical_transactions.any? %>
                       <% pairing.hcb_code.canonical_transactions.each do |ct| %>
-                        <%= render partial: "canonical_transactions/canonical_transaction", locals: { ct:, force_display_details: true, hide_tags: true, show_event_name: true } %>
+                        <%= render partial: "canonical_transactions/canonical_transaction", locals: { ct:, force_display_details: true, hide_tags: true, show_event_name: true, current_user: current_user } %>
```
Comment (YodaLightsabr) [inline app/views/static_pages/_suggested_pairings.html.erb]: Remove `current_user` locals because we've implemented a better solution with frame refreshes

File: app/views/static_pages/_suggested_pairings.html.erb [hunk 28]
```diff
@@ -1,12 +1,12 @@
...
-    <div class="card carousel__wrapper" style="overflow: unset;">
-      <div style="background-color: #ec375044; margin: -16px; padding: 16px; margin-bottom: 0px; border-radius: 0.5rem 0.5rem 0rem 0rem;">
+    <div class="card carousel__wrapper" style="overflow: unset; border: 1px solid var(--ai);">
+      <div style="margin: -16px; padding: 16px; margin-bottom: 0px; border-radius: 0.5rem 0.5rem 0rem 0rem;" class="bg-snow dark:bg-dark">
         <span tabindex="0" onclick="this.parentElement.parentElement.remove();" class="pointer">
           <%= inline_icon "view-close", size: 28, class: "pop modal__close muted  m0" %>
         </span>
         <div style="display: flex; padding-right: 32px;">
-          <h2 class="h2 mt0 mb0 pb0 border-none"><%= inline_icon "rep", size: 36, style: "margin: -8px; margin-right: 0px;" %> Suggested pairing</h2>
+          <h2 class="h2 mt0 mb0 pb0 border-none" style="color: var(--ai);"><%= inline_icon "rep", size: 36, style: "margin: -8px; margin-right: 0px;" %> Suggested pairing</h2>
```
Comment (YodaLightsabr) [inline app/views/static_pages/_suggested_pairings.html.erb]:
```suggestion
          <h2 class="h2 mt0 mb0 pb0 border-none ai"><%= inline_icon "rep", size: 36, style: "margin: -8px; margin-right: 0px;" %> Suggested pairing</h2>
```

File: app/services/receipt_service/extract.rb [hunk 29]
```diff
@@ -7,98 +7,94 @@ def initialize(receipt:)
...
 
-    def amount_cents
-      amount_cents_regex = /\$( ?[\d.,]+)(\s|\n|\\n)/
+        If you can't extract a feature, or if you can't find any features, return null for the respective keys.
+      PROMPT
 
-      amounts = match_regex(amount_cents_regex, @textual_content) { |match| match.first }
-      amounts = amounts.map do |match|
-        match[:amount] = (match[:match].to_f * 100).to_i
+      Rails.logger.level = 4
```
Comment (sampoder) [inline app/services/receipt_service/extract.rb]: We probably want to remove this line

File: app/views/hcb_codes/edit.html.erb [hunk 30]
```diff
@@ -8,19 +8,24 @@
 
 <% if @frame %>
   <%= turbo_frame_tag @hcb_code do %>
-    <%= form_with model: @hcb_code, data: { turbo: true, controller: "form" } do |form| %>
+    <%= form_with model: @hcb_code, data: { turbo: true, controller: "form" }, html: { style: "flex-grow: 1" } do |form| %>
       <div class="flex items-center">
-        <button class="pop border-none cursor-pointer mr2">
+        <button class="pop border-none cursor-pointer mr2 <%= @ai_memo ? "ai" : "" %>">
```
Comment (sampoder) [inline app/views/hcb_codes/edit.html.erb]:
```suggestion
        <button class="pop border-none cursor-pointer mr2 <%= "ai" if @ai_memo %>">
```
unimportant but kinda nice

File: app/views/receipts/_receipt.html.erb [hunk 31]
```diff
@@ -112,15 +113,59 @@
...
+      </div>
+    </div>
+
+    <div class="w-100 h-100 overflow-hidden flex md:hidden flex-col md:flex-row md:min-h min-h-min">
+      <ul class="tab-container m-0 mt-2">
+        <li class="active cursor-pointer ml-3" data-controller="scroll-into-view">
+          <a onclick="(document.querySelector('#preview_receipt_<%= instance %> .receipt-content-preview').style.display = 'block', document.querySelector('#preview_receipt_<%= instance %> .receipt-details').style.display = 'none', this.parentElement.parentElement.querySelector('.active').classList.remove('active'), this.parentElement.classList.add('active'))">Receipt</a>
+        </li>
+        <li class="cursor-pointer" data-controller="scroll-into-view">
+          <a onclick="(document.querySelector('#preview_receipt_<%= instance %> .receipt-content-preview').style.display = 'none', document.querySelector('#preview_receipt_<%= instance %> .receipt-details').style.display = 'flex', this.parentElement.parentElement.querySelector('.active').classList.remove('active'), this.parentElement.classList.add('active'))">Details</a>
```
Comment (sampoder) [inline app/views/receipts/_receipt.html.erb]: Let's make this a Stimulus controller

File: app/views/static_pages/_suggested_pairings.html.erb [hunk 32]
```diff
@@ -56,35 +56,56 @@
                 </table>
               </div>
 
+              <% if pairing.receipt.suggested_memo %>
+                <div class="p1 pl2">
+                  <%= inline_icon "rep", size: 24, style: "margin: -6px; margin-right: 0px; color: var(--ai);" %> Suggested memo rename: <strong><%= pairing.receipt.suggested_memo %></strong>
```
Comment (sampoder) [inline app/views/static_pages/_suggested_pairings.html.erb]: Thoughts on just `Suggested memo:`?

```suggestion
                  <%= inline_icon "rep", size: 24, style: "margin: -6px; margin-right: 0px; color: var(--ai);" %> Suggested memo: <strong><%= pairing.receipt.suggested_memo %></strong>
```

File: app/controllers/suggested_pairings_controller.rb [hunk 33]
```diff
@@ -34,8 +34,44 @@ def accept
...
+    @receiptable.canonical_pending_transactions.each { |cpt| cpt.update!(custom_memo:) }
+
+    respond_to do |format|
+      format.turbo_stream { render turbo_stream: generate_streams }
+      format.html         {
+        flash[:success] = { text: "Receipt linked!", link: hcb_code_path(@pairing.hcb_code), link_text: "View" }
+        redirect_back fallback_location: my_inbox_path
+      }
+    end
+  end
```
Comment (sampoder) [inline app/controllers/suggested_pairings_controller.rb]: I would combine this with `#accept` and take a `memo=true` parameter on that method that leads to it running this:

```
custom_memo = @receipt.suggested_memo
    @receiptable.canonical_transactions.each { |ct| ct.update!(custom_memo:) }
    @receiptable.canonical_pending_transactions.each { |cpt| cpt.update!(custom_memo:) }
```

File: app/views/receipts/_details.html.erb [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+  <div class="fs-mask" style="grid-template-columns: 7rem auto;">
+    <strong style="font-weight: 600;">Uploaded</strong>
+    <%= relative_timestamp receipt.created_at %>
+  </div>
+
+  <div class="fs-mask" style="grid-template-columns: 7rem auto;">
+    <strong style="font-weight: 600;">Size</strong>
+    <%= number_to_human_size receipt.file.blob.byte_size %>
+  </div>
+</section>
```
Comment (sampoder) [inline app/views/receipts/_details.html.erb]: We still seem to be having overflow issues with the file name

File: app/views/receipts/_receipt.html.erb [hunk 35]
```diff
@@ -112,15 +113,59 @@
...
+        </li>
+        <li class="cursor-pointer" data-controller="scroll-into-view">
+          <a onclick="(instance = <%= instance.to_json %>, document.querySelector(`#preview_receipt_${instance} .receipt-content-preview`).style.display = 'none', document.querySelector(`#preview_receipt_${instance} .receipt-details`).style.display = 'flex', this.parentElement.parentElement.querySelector('.active').classList.remove('active'), this.parentElement.classList.add('active'))">Details</a>
+        </li>
+        <div class="flex-grow">
+          <a href="#close_modal" tabindex="0" class="pop modal__close muted" rel="modal:close" style="background: #11111199; margin: 0px; margin-right: 6px;">
+            <%= inline_icon "view-close", size: 28, class: "pop modal__close muted  m0" %>
+          </a>
+        </div>
+      </ul>
```
Comment (sampoder) [inline app/views/receipts/_receipt.html.erb]: Let's put this in a Stimulus controller

File: app/views/static_pages/my_inbox.html.erb [hunk 36]
```diff
@@ -19,7 +19,10 @@
     <% end %>
   </p>
 
-  <%= render partial: "suggested_pairings", locals: { pairings: @pairings, current_slide: 0 } %>
+  <%= turbo_stream_from current_user, :suggested_pairings %>
+  <%= turbo_frame_tag "suggested_pairings", src: suggested_pairings_path do %>
+
+  <% end %>
```
Comment (sampoder) [inline app/views/static_pages/my_inbox.html.erb]: We probably don't want the empty block here.

It might be nice to animate in the suggested pairings though I think this is low priority

**Review by sampoder — COMMENTED** [review]
A couple of comments

**Review by sampoder — DISMISSED** [review]
LGTM; nice work on this!

**Comment by YodaLightsabr** [comment]
RSpec passes locally

**Review by maxwofford — APPROVED** [review]
Chatted through this IRL & came to the conclusion we want to flipper flag the user-facing parts of the suggestion UI to test on specific users first. I'm an approve on this going out if @garyhtou is ok with this merging while he's out.

I'm happy to be around for merge to make sure nothing goes down.
